sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -490,6 +495,101 @@ def test_spline_transformer_n_features_out(
+    is passed and missing values are present in X."""
+    X_nan = np.array([[1, 1], [2, 2], [3, 3], [np.nan, 5], [4, 4]])
+    sample_weight = [1, 1, 1, 1, 1]
+    spline = SplineTransformer(knots="quantile", handle_missing="zeros")
+
+    msg = "Passing `sample_weight` to SplineTransformer when there are"
+    with pytest.raises(NotImplementedError, match=msg):
+        spline.fit(X_nan, sample_weight=sample_weight)
+
+
+@pytest.mark.skipif(
+    sp_version < parse_version("1.8.0"),
+    reason="The option `sparse_output` is available as of scipy 1.8.0",
+)
+@pytest.mark.parametrize(
+    "extrapolation", ["error", "constant", "linear", "continue", "periodic"]
+)
+@pytest.mark.parametrize("sparse_output", [False, True])
+def test_spline_transformer_handles_missing_values(extrapolation, sparse_output):
+    """Test that SplineTransformer handles missing values correctly.
+    We only test for knots="uniform", since for "quantile" the metrics are calculated
+    differently with nans present and a different result is thus expected.
+    """
+    X_nan = np.array([[1, 1], [2, 2], [3, 3], [np.nan, 5], [4, 4]])
+    X_nan_full_column = np.array([[np.nan, np.nan], [np.nan, 1]])
+    X = np.array([[1, 1], [2, 2], [3, 3], [4, 5], [4, 4]])
+
+    # check correct error message for handle_missing="error"
+    msg = "X contains missing values (np.nan) and SplineTransformer is configured with"
+    with pytest.raises(ValueError, match=re.escape(msg)):
+        spline = SplineTransformer(
+            degree=2,
+            n_knots=3,
+            handle_missing="error",
+            extrapolation=extrapolation,
+        )
+        spline.fit_transform(X_nan)
+
+    # check correct results for handle_missing="zeros"
+    spline = SplineTransformer(
+        degree=2,
+        n_knots=3,
+        handle_missing="zeros",
+        extrapolation=extrapolation,
+        sparse_output=sparse_output,
+    )
+
+    # check for generic invariants
+    X_nan_transform = spline.fit_transform(X_nan)
+    if sparse.issparse(X_nan_transform):
+        X_nan_transform = X_nan_transform.toarray()
+    assert (X_nan_transform >= 0).all()
+    assert (X_nan_transform <= 1).all()
+
+    # check `fit_transform` does the same as `fit` and `then` transform
+    X_nan_transform = spline.fit_transform(X_nan)
+    X_nan_fit_then_transformed = spline.fit(X_nan).transform(X_nan)
+    assert_allclose_dense_sparse(X_nan_transform, X_nan_fit_then_transformed)
+
+    # check that transform works as expected when the passed data has not the same
+    # shape as the training set array:
+    X_transformed_same_shape = spline.fit_transform(X_nan)[::2]
+    X_transformed_different_shapes = spline.fit(X_nan).transform(X_nan[::2])
+    assert_allclose_dense_sparse(
+        X_transformed_same_shape, X_transformed_different_shapes
+    )
+
+    # check that if X has a feature of all nans SplineTransformer works as usual
+    spline.transform(X_nan_full_column)
```

Comment:
The result of this call should be checked to see that the first column is all zero valued.

EDIT: I see that `spline.transform(X_nan_full_column)` is called again and its output checked a few lines below.

To avoid confusion, I think the all-nan column case should be checked in a separate test function that would call `SplineTransformer` on `X_allmissing = np.asarray([[np.nan], [np.nan], [np.nan]])` for instance.

Reply:
Oh yes, I will make in into a separate test.

I'll be keeping `X_nan_full_column = np.array([[np.nan, np.nan], [np.nan, 1]])` for now. 

The reason: `X_allmissing = np.asarray([[np.nan], [np.nan], [np.nan]])` is a very special case, because in this case `np.nanmin(x)` is also nan and `BSpline.design_matrix()` raises. There is no way to avoid this, I believe, because we cannot set the existing nan values in `x` to any valid value that is within the feature space.

Reproducable:
```python
    X = np.array([[1, 1], [2, 2], [3, 3], [4, 5], [4, 4]])
    X_allmissing = np.array([[np.nan, np.nan], [np.nan, np.nan]])

    spline = SplineTransformer(
        degree=2,
        n_knots=3,
        handle_missing="zeros",
    )

    spline.fit(X)

    all_missing_column_encoded = spline.transform(X_allmissing)
```

So for now, this case fails with an error from scipy, but I think it's a quite a readable error message.